File: src/converter/plugins/convertFunctionType.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {convertParameterDeclarations} from "./convertParameterDeclaration.js";

export const convertFunctionType = createSimplePlugin((node, context, render) => {
    if (!ts.isFunctionTypeNode(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    checkCoverageService?.cover(node)

    if (node.typeParameters) {
        return `Function<Any?> /* ${typeScriptService?.printNode(node)} */`
    }

    const returnType = render(node.type)

    if (node.parameters.some(parameter => parameter.dotDotDotToken)) {
        return `Function<${returnType}> /* ${typeScriptService?.printNode(node)} */`
    }

    return convertParameterDeclarations(node, context, render, {
        strategy: "lambda",
        template: parameters => {
            return `(${parameters}) -> ${returnType}`
        },
    })
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertFunctionType.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.asArray
import typescript.isFunctionTypeNode

val convertFunctionType = createPlugin plugin@{ node, context, render ->
    if (!isFunctionTypeNode(node)) return@plugin null

    val checkCoverageService = context.lookupService(checkCoverageServiceKey)
    val typeScriptService = context.lookupService(typeScriptServiceKey)

    checkCoverageService?.cover(node)

    if (node.typeParameters != null) {
        return@plugin "Function<Any?> /* ${typeScriptService?.printNode(node)} */"
    }

    val returnType = render(node.type)

    if (node.parameters.asArray().any { it.dotDotDotToken != null }) {
        return@plugin "Function<${returnType}> /* ${typeScriptService?.printNode(node)} */"
    }

    convertParameterDeclarations(node, context, render, ParameterDeclarationsConfiguration(
        strategy = ParameterDeclarationStrategy.lambda,
        template = { parameters, _ ->
            "(${parameters}) -> $returnType"
        },
    ))
}
